README
                         
			       EEGVIEW


Presentation
============

eegview is a small program to display/record EEG signals. It is particularly
useful for setting up quickly the acquisition part of a offline protocol.


Supported platforms
===================

Any platforms supported by eegdev, mcpanel and xdffileio.


Compilation
===========

This library is organized as a GNU package and can be compiled and
installed in the same way (see INSTALL file for further information).



NEWS

eegview-0.0
===========

Initial release


main.c
#include <stdio.h>
#include <eegpanel.h>
#include <act2demux.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <bdffile.h>


/**************************************************************************
 *                                                                        *
 *              Channel selection structures                              *
 *                                                                        * 
 **************************************************************************/
struct channel_option {
	int type;
	const char* string;
	int numch;
};

struct channel_option eeg_options[] = {
	{ALLEEG, "all", -1},
	{A, "A", 32},
	{AB, "AB", 64},
	{AD, "AD", 128}
};
#define NUM_EEG_OPTION (sizeof(eeg_options)/sizeof(eeg_options[0]))

struct channel_option exg_options[] = {
	{ALLEXG, "all", -1},
	{STD, "STD", 8},
	{NONE, "none", 0}
};
#define NUM_EXG_OPTION (sizeof(exg_options)/sizeof(exg_options[0]))

const struct channel_option* eeg_opt = eeg_options+2;
const struct channel_option* exg_opt = eeg_options+1;



/**************************************************************************
 *                                                                        *
 *              Global variables                                          *
 *                                                                        * 
 **************************************************************************/
pthread_t thread_id;
BDFFile* bdffile = NULL;
volatile int run_eeg = 0;
volatile int record_file = 0;
#define NSAMPLES	32
ActivetwoSystemInfo info;

/**************************************************************************
 *                                                                        *
 *              Error message helper functions                            *
 *                                                                        * 
 **************************************************************************/
static char* const  acquisition_system_message[] = {
	"Attempt to open the driver failed.\n\nCheck that it is correctly installed, or that the USB cable is plugged.",
	"The synchronization with the ADC box is lost.\n\nCheck that the ADC box is switched on or that the optical cable is plugged.",
	"An attempt to get the data in buffered mode has been made, but the acquisition has not been started yet.",
	"The buffer is full. Data has been lost",
	"The specified option is not supported or is not consistent with the capabilities of the system"
};
#define NUM_SYS_MESSAGE	(sizeof(acquisition_system_message)/sizeof(char*))

const char* get_acq_message(ActivetwoRetCode retcode)
{
	int msg_ind  = SUCCESS - retcode -1;

	if (msg_ind > NUM_SYS_MESSAGE)
		return "Unknown error.";
	return acquisition_system_message[SUCCESS - retcode -1];
}

static char bdffile_message[128];
const char* get_bdf_message(void)
{
	sprintf(bdffile_message,"BDF Error (%i)",BDFGetLastError());

	return bdffile_message;
}

/**************************************************************************
 *                                                                        *
 *              Acquition system callbacks                                *
 *                                                                        * 
 **************************************************************************/
// EEG acquisition thread
void* reading_thread(void* arg)
{
	float *eeg, *exg;
	uint32_t *tri;
	int32_t *raweeg, *rawexg;
	EEGPanel* panel = arg;
	unsigned int neeg, nexg;

	neeg = eeg_opt->numch;
	nexg = exg_opt->numch;

	eeg = calloc(neeg*NSAMPLES, sizeof(*eeg));
	exg = calloc(nexg*NSAMPLES, sizeof(*exg));
	tri = calloc(NSAMPLES, sizeof(*tri));
	raweeg = (int32_t*)eeg;
	rawexg = (int32_t*)exg;
	
	ActivetwoStartBufferedAcquisition();
	while(run_eeg) {
		if (ActivetwoGetBufferedSamples(NSAMPLES, tri, raweeg, rawexg)==SUCCESS) {
			Act2ScaleArray(float, eeg, int32_t, raweeg, NSAMPLES*neeg, ACTIVE_ELEC_SCALE);
			Act2ScaleArray(float, exg, int32_t, rawexg, NSAMPLES*nexg, ACTIVE_ELEC_SCALE);
			Act2ScaleArray(uint32_t, tri, uint32_t, tri, NSAMPLES, TRIGGER_SCALE);
		
			eegpanel_add_samples(panel, eeg, exg, tri, NSAMPLES);
		}
	}
	ActivetwoStopBufferedAcquisition();

	free(eeg);
	free(exg);
	free(tri);

	return 0;
}

// Connection to the system 
int Connect(EEGPanel* panel)
{
	int retval;
	
	if ((retval = ActivetwoConnectToSystem()) < 0)
		return retval;

	// Set the number of channels of the "All channels" values
	info = ActivetwoGetSystemInfo();
	eeg_options[0].numch = info.maxnumber_eeg_channels;
	exg_options[0].numch = info.maxnumber_sensor_channels;
	

	// Set the acquisition according to the settings
	if ((retval = ActivetwoSetDataFormats(eeg_opt->numch, exg_opt->numch))!=SUCCESS) {
		ActivetwoDisconnectFromSystem();
		return retval;
	}
	eegpanel_define_input(panel, eeg_opt->numch, exg_opt->numch, 16, info.samplerate);

	run_eeg = 1;
	pthread_create(&thread_id, NULL, reading_thread, panel);
	
	return 0;
}


int Disconnect(EEGPanel* panel)
{
	run_eeg = 0;
	pthread_join(thread_id, NULL);
	ActivetwoDisconnectFromSystem();
	return 0;
}


int SystemConnection(int start, void* user_data)
{
	EEGPanel* panel = user_data;
	int retval;

	retval = start ? Connect(panel) : Disconnect(panel);
	if (retval)
		eegpanel_popup_message(panel, get_acq_message(retval));

	return (retval < 0) ? 0 : 1;
}

/**************************************************************************
 *                                                                        *
 *              File recording callbacks                                  *
 *                                                                        * 
 **************************************************************************/
int SetupRecording(const ChannelSelection * eeg_sel,
		   const ChannelSelection * exg_sel, void *user_data)
{
	EEGPanel *panel = user_data;
	char *filename;
	char tmpstr[64];
	int j;
	BDFChannel chann;

	filename =
	    eegpanel_open_filename_dialog(panel, "*.bdf",
					  "BDF files (*.bdf)");
	
	// Check that user hasn't press cancel
	if (!filename)
		return 0;

	// Create the BDF file
	bdffile = OpenBDFFile(filename, BDF_MODE_WRITE);
	if (!bdffile) 
		goto abort;


	// Set up the channels
	SetSamplingRateBDFFile(bdffile, info.samplerate);
	for (j = 0; j < eeg_opt->numch; j++) {
		sprintf(tmpstr, "EEG%i", j);
		SetBDFChannel(&chann, BDF_CHANNEL_EEG, tmpstr);
		if (AddBDFChannel(bdffile, &chann) < 1)
			goto abort;
	}
	for (j = 0; j < eeg_opt->numch; j++) {
		sprintf(tmpstr, "EXG%i", j);
		SetBDFChannel(&chann, BDF_CHANNEL_EXG, tmpstr);
		if (AddBDFChannel(bdffile, &chann) < 1)
			goto abort;
	}
	SetBDFChannel(&chann, BDF_CHANNEL_TRIGG, "Status");
	if (AddBDFChannel(bdffile, &chann) < 1)
		goto abort;

	// Make the file ready for recording
	if(PrepareForTransfer(bdffile))
		goto abort;

	return 1;
	
abort:
	eegpanel_popup_message(panel, get_bdf_message());
	CloseBDFFile(bdffile);
	return 0;
}



/**************************************************************************
 *                                                                        *
 *              Initialization of the application                         *
 *                                                                        * 
 **************************************************************************/
const char* opt_str[] = {
	"--settings=",
	"--ui-file=",
	"--map-file=",
	"--eeg-set=",
	"--exg-set="
};
#define NUM_OPTS	(sizeof(opt_str)/sizeof(opt_str[0]))

int main(int argc, char* argv[])
{
	EEGPanel* panel;
	int i, iopt;
	int retval = 0;
	const char* opt_vals[NUM_OPTS] = {NULL};

	// Process command line options
	init_eegpanel_lib(&argc, &argv);
	for (i=1; i<argc; i++) {
		for (iopt=0; iopt<NUM_OPTS; iopt++) {
			if (strncmp(opt_str[iopt],argv[i],strlen(opt_str[iopt]))==0) {
				opt_vals[iopt] = argv[i]+strlen(opt_str[iopt]);
				break;
			}
		}

		if (strcmp(argv[i],"--version")==0) {
			printf("%s: version ??? build on\n\t%s\n\t%s\n",argv[0],ActivetwoGetString(),BDFFileGetString());
			return 0;
		}

		if ( (iopt == NUM_OPTS)
		     && (strcmp(argv[i], "--help")!=0)
		     && (strcmp(argv[i], "-h")!=0) ) {
		     	fprintf(stderr, "Unknown option\n");
			retval = 1;
		}

		printf("Usage: %s [GTK+ OPTIONS...] [--settings=FILE] [--map-file=FILE] [--ui-file=FILE] [--eeg-set=EEG_SET] [--exg-set=EXG_SET] [--version]\n",argv[0]);

		return retval;
	}
	

	

	panel = eegpanel_create(opt_vals[1], opt_vals[0]);
	if (!panel) {
		fprintf(stderr,"error at the creation of the panel\n");
		return 2;
	}
	
	// Setup callbacks
	panel->user_data = panel;
	panel->system_connection = SystemConnection;
	panel->setup_recording = SetupRecording;

	// Run the panel
	eegpanel_show(panel, 1);
	eegpanel_run(panel, 0);


	eegpanel_destroy(panel);

	return 0;
}



src/fakeact2.c
#include <act2demux.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static int64_t ns_read;
static struct timespec start;


static unsigned int neeg, nexg;
static ActivetwoSystemInfo info = 
{
	.samplerate = 2048,
	.maxnumber_eeg_channels = 64,
	.maxnumber_sensor_channels = 24,
	.mk_model = 2,
	.speedmode = 4
};

ActivetwoRetCode ActivetwoConnectToSystem(void)
{
	return SUCCESS;
}

void ActivetwoDisconnectFromSystem(void)
{
	
}

ActivetwoSystemInfo ActivetwoGetSystemInfo(void)
{
	return info;
}

ActivetwoRetCode ActivetwoSetDataFormats(EEGformat eeg_format, EXGformat exg_format)
{
	if (eeg_format == A)
		neeg = 32;
	else if (eeg_format == AB)
		neeg = 64;
	else if (eeg_format == ALLEEG)
		neeg = info.maxnumber_eeg_channels;
	else
		return BAD_OPTION;

	if (exg_format == NONE)
		nexg = 0;
	else if (exg_format == STD)
		nexg = 8;
	else if (exg_format == ALLEXG)
		nexg = info.maxnumber_sensor_channels;
	else
		return BAD_OPTION;

	return SUCCESS;
}

ActivetwoRetCode ActivetwoGetLastSamples(unsigned int nSamples, uint32_t* triggers, int32_t* eeg, int32_t* sensors)
{
	return SUCCESS;
}

ActivetwoRetCode ActivetwoStartBufferedAcquisition()
{
	ns_read = 0;
	clock_gettime(CLOCK_MONOTONIC, &start);
	return SUCCESS;
}

void ActivetwoStopBufferedAcquisition()
{
	
}

ActivetwoRetCode ActivetwoGetBufferedSamples(unsigned int nSamples, uint32_t* triggers, int32_t* eeg, int32_t* sensors)
{
	int ns_unread;
	int i, seed;
	uint32_t val;
	for (i=0; i<nSamples; i++) {
		val = ((ns_read + i)/128) % 8;
		triggers[i] = val << 8;
	}
	//memset(triggers, 0, nSamples*sizeof(uint32_t));
	

	memset(eeg, 0, nSamples*neeg*sizeof(int32_t));
	memset(sensors, 0, nSamples*nexg*sizeof(int32_t));

	while (ActiveTwoGetNumUnreadBuffered() < nSamples)
		usleep(10);

	ns_read += nSamples;


	return SUCCESS;
}

int ActiveTwoGetNumUnreadBuffered(void)
{
	struct timespec curr;
	int64_t ns_acq;
	double seconds;

	clock_gettime(CLOCK_MONOTONIC, &curr);

	seconds = (double)(curr.tv_sec - start.tv_sec)
		  + 1e-9*(double)(curr.tv_nsec - start.tv_nsec);

	ns_acq = (int64_t)(seconds * (double)info.samplerate);


	return ns_acq - ns_read;
}


src/configuration.c
#include "configuration.h"
#include <eegpanel.h>
#include <confuse.h>
#include <stdlib.h>
#include <locale.h>
#include <string.h>

// Settings for the allowed "vacabulary" for configuration file
cfg_opt_t filt_opts[] = {
	CFG_BOOL("state", cfg_false, CFGF_NONE),
	CFG_FLOAT("freq", 0.0, CFGF_NONE),
	CFG_END()
};
cfg_opt_t set_opts[] = {
	CFG_STR_LIST("labels", "{}", CFGF_NONE),
	CFG_END()
};

cfg_opt_t opts[] = {
	CFG_SEC("filter", filt_opts, CFGF_MULTI | CFGF_TITLE), 
	CFG_SEC("map", set_opts, CFGF_MULTI | CFGF_TITLE), 
	CFG_FUNC("include", cfg_include),
	CFG_END()
};

void free_label_list(char** labels)
{
	unsigned int i;

	if (!labels)
		return;

	i = 0;
	while (labels[i])
		free(labels[i++]);

	free(labels);
}

int get_filter_settings(cfg_t *cfg, struct FilterSettings* filtset)
{
	if (cfg == NULL)
		return 0;

	filtset->state = cfg_getbool(cfg, "state") == cfg_true ? 1 : 0;
	filtset->freq = cfg_getfloat(cfg, "freq");
	return 0;
}

int get_set_labels(cfg_t *cfg, char*** label_list)
{
	unsigned int nchann, i;
	char *value, **labels;

	if (cfg == NULL)
		return 0;
	
	// Allocate the list
	nchann = cfg_size(cfg, "labels");
	labels = calloc(nchann+1, sizeof(char*));
	if (labels == NULL)
		return 0;
	
	// Allocate and copy each labels
	for (i=0; i<nchann; i++) {
		value = cfg_getnstr(cfg, "labels", i);
		labels[i] = malloc(strlen(value)+1);
		if (!labels[i]) {
			free_label_list(labels);
			return 0;
		}
		strcpy(labels[i], value);
	}

	*label_list = labels;
	return nchann;
}

int read_configuration(struct PanelSettings* settings, const char* eegset, const char* sensorset, const char* filename)
{
	const char* setname;
	int state;
	double freq;
	unsigned int nchann;
	char** labels = NULL;
	cfg_t *cfg = NULL;

	
	setlocale(LC_NUMERIC, "POSIX");

	cfg = cfg_init(opts, 0);
	cfg_parse(cfg, filename);

	// Read the parsed information related to filters
	get_filter_settings(cfg_gettsec(cfg, "filter", "eeglow"), &(settings->filt[EEG_LOWPASS_FLT]));
	get_filter_settings(cfg_gettsec(cfg, "filter", "sensorlow"), &(settings->filt[SENSOR_LOWPASS_FLT]));
	get_filter_settings(cfg_gettsec(cfg, "filter", "eeghigh"), &(settings->filt[EEG_HIGHPASS_FLT]));
	get_filter_settings(cfg_gettsec(cfg, "filter", "sensorhigh"), &(settings->filt[SENSOR_HIGHPASS_FLT]));

	// Read the parsed information related to labels
	nchann = get_set_labels(cfg_gettsec(cfg, "map", eegset), &labels);
	if (labels) {
		settings->eeglabels = (const char**)labels;
		settings->num_eeg = nchann;
	}
	nchann = get_set_labels(cfg_gettsec(cfg, "map", sensorset), &labels);
	if (labels) {
		settings->sensorlabels = (const char**)labels;
		settings->num_sensor = nchann;
	}
	
	cfg_free(cfg);

	return 0;
}

void free_configuration(struct PanelSettings* settings)
{
	free_label_list((char**)settings->eeglabels);
	free_label_list((char**)settings->sensorlabels);
}


src/eegview.c
/*
    Copyright (C) 2009-2011  EPFL (Ecole Polytechnique Fédérale de Lausanne)
    Laboratory CNBI (Chair in Non-Invasive Brain-Machine Interface)
    Nicolas Bourdaud <nicolas.bourdaud@gmail.com>

    This program is free software: you can redistribute it and/or modify
    modify it under the terms of the version 3 of the GNU General Public
    License as published by the Free Software Foundation.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <stdio.h>
#include <mcpanel.h>
#include <eegdev.h>
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include <xdfio.h>
#include <errno.h>
#include <getopt.h>

#if HAVE_CONFIG_H
# include <config.h>
#endif

/**************************************************************************
 *                                                                        *
 *              Global variables                                          *
 *                                                                        * 
 **************************************************************************/
static const char* uifilename = NULL;
static const char* devstring = NULL;

pthread_t thread_id;
pthread_mutex_t sync_mtx = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t file_mtx = PTHREAD_MUTEX_INITIALIZER;
struct eegdev* dev = NULL;
struct xdf* xdf = NULL;
int run_eeg = 0;
int record_file = 0;
#define NSAMPLES	32

size_t strides[3];
struct grpconf grp[] = {
	{
		.sensortype = EGD_EEG,
		.index = 0,
		.iarray = 0,
		.arr_offset = 0,
		.datatype = EGD_FLOAT
	},
	{
		.sensortype = EGD_SENSOR,
		.index = 0,
		.iarray = 1,
		.arr_offset = 0,
		.datatype = EGD_FLOAT
	},
	{
		.sensortype = EGD_TRIGGER,
		.index = 0,
		.iarray = 2,
		.arr_offset = 0,
		.datatype = EGD_INT32
	}
};
	
static char **labels[2] = {NULL, NULL};

#define NSCALE 2
static const char* scale_labels[NSCALE] = {"25.0mV", "50.0mV"};
static const float scale_values[NSCALE] = {25.0e3, 50.0e3};

#define NTAB 3
static struct panel_tabconf tabconf[NTAB] = {
	{.type = TABTYPE_SCOPE, .name = "EEG"},
	{.type = TABTYPE_BARGRAPH, .name = "Offsets", .nscales = NSCALE,
	 .sclabels = scale_labels, .scales = scale_values},
	{.type = TABTYPE_SCOPE, .name = "Sensors"}
};

static int StopRecording(void* user_data);
/**************************************************************************
 *                                                                        *
 *              Error message helper functions                            *
 *                                                                        * 
 **************************************************************************/
const char* get_acq_msg(int error)
{
	return strerror(error);
}

static char bdffile_message[128];
/**************************************************************************
 *                                                                        *
 *              Acquition system callbacks                                *
 *                                                                        * 
 **************************************************************************/
static
void free_labels(void)
{
	unsigned int igrp, i;

	for (igrp=0; igrp<2; igrp++) {
		if (labels[igrp] == NULL)
			continue;
		for (i=0; i<grp[igrp].nch; i++) 
			free(labels[igrp][i]);

		free(labels[igrp]);
		labels[igrp] = NULL;
	}
}

static
int get_labels_from_device(void)
{
	unsigned int i, igrp;
	int type;

	// Allocate and copy each labels
	for (igrp=0; igrp<2; igrp++) {
		labels[igrp] = malloc(grp[igrp].nch * sizeof(char*));
		type = grp[igrp].sensortype;
		for (i=0; i<grp[igrp].nch; i++) {
			labels[igrp][i] = malloc(32);
			egd_channel_info(dev, type, i,
			               EGD_LABEL, labels[igrp][i], EGD_EOL);
		}
	}
	return 0;
}


static
void* display_bdf_error(void* arg)
{
	mcpanel* pan = arg;
	mcp_popup_message(pan, bdffile_message);
	return NULL;
}

static
int device_connection(void)
{
	int retval;
	if (!(dev = egd_open(devstring)))
		return errno;

	// Set the number of channels of the "All channels" values
	grp[0].nch = egd_get_numch(dev, EGD_EEG);
	grp[1].nch = egd_get_numch(dev, EGD_SENSOR);
	grp[2].nch = egd_get_numch(dev, EGD_TRIGGER) ? 1 : 0;
	strides[0] = grp[0].nch * sizeof(float);
	strides[1] = grp[1].nch * sizeof(float);
	strides[2] = grp[2].nch * sizeof(int32_t);

	get_labels_from_device();

	// Set the acquisition according to the settings
	if (egd_acq_setup(dev, 3, strides, 3, grp)) {
		retval = errno;
		egd_close(dev);
		return retval;
	}

	return 0;
}


static
int device_disconnection(void)
{
	free_labels();
	egd_close(dev);
	return 0;
}


// EEG acquisition thread
static
void* reading_thread(void* arg)
{
	float *eeg, *exg;
	int32_t *tri;
	mcpanel* panel = arg;
	unsigned int neeg, nexg;
	int run_acq, error, saving = 0;
	ssize_t nsread;

	neeg = grp[0].nch;
	nexg = grp[1].nch;

	eeg = neeg ? calloc(neeg*NSAMPLES, sizeof(*eeg)) : NULL;
	exg = nexg ? calloc(nexg*NSAMPLES, sizeof(*exg)) : NULL;
	tri = calloc(NSAMPLES, sizeof(*tri));
	
	egd_start(dev);
	while (1) {
		
		// update control flags
		pthread_mutex_lock(&sync_mtx);
		run_acq = run_eeg;
		if (saving != record_file) {
			// report write status back
			if (record_file)
				pthread_mutex_lock(&file_mtx);
			else
				pthread_mutex_unlock(&file_mtx);
			saving = record_file;
		}
		pthread_mutex_unlock(&sync_mtx);

		// Check the stop acquisition flag
		if (!run_acq)
			break;

		// Get data from the system
		nsread = egd_get_data(dev, NSAMPLES, eeg, exg, tri);
		if (nsread < 0) {
			error = errno;			
			mcp_notify(panel, DISCONNECTED);
			mcp_popup_message(panel, get_acq_msg(error));
			break;
		}

		
		// Write samples on file
		if (saving) {
			if (xdf_write(xdf, nsread, eeg, exg, tri) < 0) {
				pthread_attr_t attr;
				pthread_t thid;
				sprintf(bdffile_message,"XDF Error: %s",strerror(errno));
			
				// Stop recording
				saving = 0;
				pthread_mutex_unlock(&file_mtx);
				StopRecording(NULL);

				// Pop up message
				pthread_attr_init(&attr);
				pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
				pthread_create(&thid, &attr, display_bdf_error, panel);
				pthread_attr_destroy(&attr);
			}
		}

		mcp_add_samples(panel, 0, nsread, eeg);
		mcp_add_samples(panel, 1, nsread, eeg);
		mcp_add_samples(panel, 2, nsread, exg);
		mcp_add_triggers(panel, nsread, (const uint32_t*)tri);
	}

	if (saving)
		pthread_mutex_unlock(&file_mtx);

	egd_stop(dev);

	free(eeg);
	free(exg);
	free(tri);

	return 0;
}

// Connection to the system 
static
int Connect(mcpanel* panel)
{
	int retval;
	const char*** clabels;
	float fs;

	retval = device_connection();
	if (retval)
		return retval;

	fs = egd_get_cap(dev, EGD_CAP_FS, NULL);
	clabels = (const char***)labels;

	// Setup the panel with the settings
	mcp_define_tab_input(panel, 0, grp[0].nch, fs, clabels[0]);
	mcp_define_tab_input(panel, 1, grp[0].nch, fs, clabels[0]);
	mcp_define_tab_input(panel, 2, grp[1].nch, fs, clabels[1]);
	mcp_define_triggers(panel, 16, fs);

	pthread_mutex_lock(&sync_mtx);
	run_eeg = 1;
	pthread_mutex_unlock(&sync_mtx);
	pthread_create(&thread_id, NULL, reading_thread, panel);
	
	return 0;
}


static
int Disconnect(mcpanel* panel)
{
	(void)panel;

	pthread_mutex_lock(&sync_mtx);
	run_eeg = 0;
	pthread_mutex_unlock(&sync_mtx);

	pthread_join(thread_id, NULL);
	device_disconnection();
	return 0;
}


static
int SystemConnection(int start, void* user_data)
{
	mcpanel* panel = user_data;
	int retval;

	retval = start ? Connect(panel) : Disconnect(panel);
	if (retval) {
		mcp_popup_message(panel, get_acq_msg(retval));
		return 0;
	}

	return 1;
}


static char devinfo[1024];

static
void on_device_info(int id, void* data)
{
	(void)id;
	unsigned int sampling_freq, eeg_nmax, sensor_nmax, trigger_nmax;
	char *device_type, *device_id;
	char prefiltering[128];
	mcpanel* panel = data;

	if (!run_eeg)
		return;

	egd_get_cap(dev, EGD_CAP_DEVTYPE, &device_type);
	egd_get_cap(dev, EGD_CAP_DEVID, &device_id);
	egd_get_cap(dev, EGD_CAP_FS, &sampling_freq);
	eeg_nmax = egd_get_numch(dev, EGD_EEG);
	sensor_nmax = egd_get_numch(dev, EGD_SENSOR);
	trigger_nmax = egd_get_numch(dev, EGD_TRIGGER);
	egd_channel_info(dev, EGD_EEG, 0,
			EGD_PREFILTERING, prefiltering, EGD_EOL);
	
	snprintf(devinfo, sizeof(devinfo)-1,
	       "system info:\n\n"
	       "device type: %s\n"
	       "device model: %s\n"
	       "sampling frequency: %u Hz\n"
	       "num EEG channels: %u\n"
	       "num sensor channels: %u\n"
	       "num trigger channels: %u\n"
	       "prefiltering: %s\n",
	       device_type, device_id, sampling_freq,
	       eeg_nmax, sensor_nmax, trigger_nmax, prefiltering);
	
	mcp_popup_message(panel, devinfo);	
}

/**************************************************************************
 *                                                                        *
 *              File recording callbacks                                  *
 *                                                                        * 
 **************************************************************************/
static
int setup_xdf_channel_group(int igrp)
{
	char label[32], transducter[128], unit[16], filtering[128];
	double mm[2];
	unsigned int j;
	int isint;
	struct xdfch* ch;

	egd_channel_info(dev, grp[igrp].sensortype, 0,
			 EGD_UNIT, unit,
			 EGD_TRANSDUCTER, transducter,
			 EGD_PREFILTERING, filtering,
			 EGD_MM_D, mm,
			 EGD_ISINT, &isint,
			 EGD_EOL);

	xdf_set_conf(xdf, 
	               XDF_CF_ARRINDEX, igrp,
		       XDF_CF_ARROFFSET, 0,
		       XDF_CF_ARRDIGITAL, 0,
		       XDF_CF_ARRTYPE, isint ? XDFINT32 : XDFFLOAT,
		       XDF_CF_PMIN, mm[0],
		       XDF_CF_PMAX, mm[1],
		       XDF_CF_TRANSDUCTER, transducter,
	               XDF_CF_PREFILTERING, filtering,
		       XDF_CF_UNIT, unit,
		       XDF_NOF);

	for (j = 0; j < grp[igrp].nch; j++) {
		egd_channel_info(dev, grp[igrp].sensortype, j,
		                 EGD_LABEL, label, EGD_EOL);

		// Add the channel to the BDF
		if ((ch = xdf_add_channel(xdf, label)) == NULL)
			return -1;
	}
	return 0;
}

static
int SetupRecording(void *user_data)
{
	mcpanel *panel = user_data;
	char *filename;
	unsigned int j;
	int fs = egd_get_cap(dev, EGD_CAP_FS, NULL);

	filename = mcp_open_filename_dialog(panel,
	                              "BDF files|*.bdf|*.BDF||Any files|*");
	
	// Check that user hasn't press cancel
	if (filename == NULL)
		return 0;

	// Create the BDF file
	xdf = xdf_open(filename, XDF_WRITE, XDF_BDF);
	if (!xdf) 
		goto abort;

	// Configuration file genral header
	xdf_set_conf(xdf,
	             XDF_F_REC_DURATION, 1.0,
	             XDF_F_REC_NSAMPLE, fs,
		     XDF_NOF);

	// Set up the channels
	for (j=0; j<3; j++)	
		if (setup_xdf_channel_group(j))
			goto abort;

	// Make the file ready for recording
	xdf_define_arrays(xdf, 3, strides);
	if (xdf_prepare_transfer(xdf))
		goto abort;

	return 1;
	
abort:
	sprintf(bdffile_message,"XDF Error: %s",strerror(errno));
	mcp_popup_message(panel, bdffile_message);
	xdf_close(xdf);
	return 0;
}

static
int StopRecording(void* user_data)
{
	(void)user_data;

	pthread_mutex_lock(&sync_mtx);
	record_file = 0;
	pthread_mutex_unlock(&sync_mtx);
	
	pthread_mutex_lock(&file_mtx);
	xdf_close(xdf);
	xdf = NULL;
	pthread_mutex_unlock(&file_mtx);

	return 1;
}


static
int ToggleRecording(int start, void* user_data)
{
	(void)user_data;
	pthread_mutex_lock(&sync_mtx);
	record_file = start;
	pthread_mutex_unlock(&sync_mtx);
	return 1;
}



/**************************************************************************
 *                                                                        *
 *              Initialization of the application                         *
 *                                                                        * 
 **************************************************************************/
enum option_index {
	UIFILE,
	DEVSTRING,
	SOFTWAREVERSION,
	HELP,
	NUM_OPTS
};

static struct option opt_str[] = {
	[UIFILE] = {"ui-file", 1, NULL, 0},
	[DEVSTRING] = {"device", 1, NULL, 0},
	[SOFTWAREVERSION] = {"version", 0, NULL, 0},
	[HELP] = {"help", 0, NULL, 'h'}

};


static void print_usage(const char* cmd)
{
	fprintf(stdout,
"Usage: %s [GTK+ OPTIONS...]\n"
"            [--device=DEVSTRING] [--ui-file=FILE]\n"
"            [--version] [--help | -h]\n",
               cmd);
	
}

static void print_version(void)
{
	printf("%s: version %s build on\n\t%s\n\t%s\n",
	       PACKAGE_NAME, 
	       PACKAGE_VERSION,
	       egd_get_string(),
	       xdf_get_string());
}

static 
int process_options(int argc, char* argv[])
{
	int c;
	int option_index = 0;

	while (1) {
		c = getopt_long(argc, argv, "h", opt_str, &option_index);
		if (c == -1)
			break;

		switch (c) {
		case 0:
			if (option_index == UIFILE)
				uifilename = optarg;
			else if (option_index == DEVSTRING)
				devstring = optarg;
			else if (option_index == SOFTWAREVERSION)
				print_version();
			break;

		case 'h':
			print_usage(argv[0]);
			return 1;

		case '?':
			print_usage(argv[0]);
			return -1;
		}
	}

	return 0;
}


int main(int argc, char* argv[])
{
	mcpanel* panel = NULL;
	int retval = 0, retcode = EXIT_FAILURE;
	struct panel_button custom_button = {
		.label = "device info",
		.id = 0,
		.callback = on_device_info
	};
	struct PanelCb cb = {
		.user_data = NULL,
		.system_connection = SystemConnection,
		.setup_recording = SetupRecording,
		.stop_recording = StopRecording,
		.toggle_recording = ToggleRecording,
		.nbutton = 1,
		.custom_button = &custom_button,
		.confname = PACKAGE_NAME
	};

	// Process command line options
	mcp_init_lib(&argc, &argv);
	retval = process_options(argc, argv);
	if (retval)
		return (retval > 0) ? 0 : -retval;
	

	panel = mcp_create(uifilename, &cb, NTAB, tabconf);
	if (!panel) {
		fprintf(stderr,"error at the creation of the panel\n");
		goto exit;
	}
	
	// Run the panel
	mcp_show(panel, 1);
	mcp_run(panel, 0);
	if (run_eeg)
		Disconnect(panel);

	mcp_destroy(panel);
	retcode = EXIT_SUCCESS;

exit:
	return retcode;
}

